src/stan/mcmc/mpi_auto_adaptation.hpp

```diff
@@ -0,0 +1,285 @@
+      Eigen::VectorXd v = initial_guess;
+      double eval = 0.0;
+      Eigen::VectorXd Av = f(v);
+      stan::math::check_matching_sizes("power_method", "matrix vector product", Av, "vector", v);
+
+      int i = 0;
+      for(; i < max_iterations; ++i) {
+	double v_norm = v.norm();
+	double new_eval = v.dot(Av) / (v_norm * v_norm);
+	if(i == max_iterations - 1 || std::abs(new_eval - eval) <= tol * std::abs(eval)) {
+	  tol = std::abs(new_eval - eval) / std::abs(eval);
+	  eval = new_eval;
+	  max_iterations = i + 1;
+	  break;
+	}
+
+	eval = new_eval;
+	v = Av / Av.norm();
+
+	Av = f(v);
+      }
+
+      return eval;
+    }
+
+    /**
+     * Compute the largest eigenvalue of the sample covariance rescaled by a metric,
+     *  that is, the largest eigenvalue of L^{-1} \Sigma L^{-T}
+     *
+     * @param L Cholesky decomposition of Metric
+     * @param Sigma Sample covariance
+     * @return Largest eigenvalue
+     */
+    double eigenvalue_scaled_covariance(const Eigen::MatrixXd& L, const Eigen::MatrixXd& Sigma) {
+      Eigen::MatrixXd S = L.template triangularView<Eigen::Lower>().
+	solve(L.template triangularView<Eigen::Lower>().solve(Sigma).transpose()).transpose();
+
+      auto Sx = [&](Eigen::VectorXd x) -> Eigen::VectorXd {
+	return S * x;
+      };
+
+      int max_iterations = 100;
+      double tol = 1e-3;
+
+      return internal::power_method(Sx, Eigen::VectorXd::Random(Sigma.cols()), max_iterations, tol);
+    }
+
+    /**
+     * Compute the largest eigenvalue of the Hessian of the log density rescaled by a metric,
+     *  that is, the largest eigenvalue of L^T \nabla^2_{qq} H(q) L
+     *
+     * @tparam Model Type of model
+     * @param model Defines the log density
+     * @param q Point around which to compute the Hessian
+     * @param L Cholesky decomposition of Metric
+     * @return Largest eigenvalue
+     */
+    template<typename Model>
+    double eigenvalue_scaled_hessian(const Model& model, const Eigen::MatrixXd& L, const Eigen::VectorXd& q) {
+      Eigen::VectorXd eigenvalues;
+      Eigen::MatrixXd eigenvectors;
+
+      auto hessian_vector = [&](const Eigen::VectorXd& x) -> Eigen::VectorXd {
+	double lp;
+	Eigen::VectorXd grad1;
+	Eigen::VectorXd grad2;
+	//stan::math::hessian_times_vector(log_prob_wrapper_covar<Model>(model), q, x, lp, Ax);
+	double dx = 1e-5;
+	Eigen::VectorXd dr = L * x * dx;
+	stan::math::gradient(log_prob_wrapper_covar<Model>(model), q + dr / 2.0, lp, grad1);
+	stan::math::gradient(log_prob_wrapper_covar<Model>(model), q - dr / 2.0, lp, grad2);
+	return L.transpose() * (grad1 - grad2) / dx;
+      };
+ 
+      int max_iterations = 100;
+      double tol = 1e-3;
+	
+      return internal::power_method(hessian_vector, Eigen::VectorXd::Random(q.size()), max_iterations, tol);
+    }
+  }
+
+template <typename Model>
+class mpi_auto_adaptation : public mpi_metric_adaptation {
+#ifdef STAN_LANG_MPI
+  using est_t = stan::math::mpi::mpi_covar_estimator;
```

Comment:
I've updated the `dense_e` 's `covar` calculation and no longer use the naive version of `mpi_covar_estimator`. It's intended to improve communication performance. It's ok for you to use the old version.